enhancements/image-references.md

```diff
@@ -0,0 +1,234 @@
+sources as input to the creation process.
+
+### Proposed UX
+
+A set of new files in the `metadata/` directory of a bundle will be authored by
+the manifest authors and build systems to record image replacement intent and
+execution.
+
+Operator manifest authors write a manifest that refers to images using some
+opaque string, and provide an `image-references.yaml` file alongside their
+manifests that declares which strings inside of their manifest are referring
+to pull specifications of images and names each occurence.
+
+The `metadata/image-references.yaml` file holds data in the following format, mapping
+common names of container images to their string placeholders in the manifests:
+
+```yaml
+imageReferences:
+- name: common-name
+  substitute: registry.svc.ci.openshift.org/openshift:image
+```
+
+This mapping, therefore, defines what needs to be replaced in manifests when
+they are bundled and identifies each replacement with a name. When
+`operator-sdk generate bundle` runs, it will require as input a second mapping
+from those names to literal image pull specifications and will run the replacement.
+In this manner, the configuration provided by the manifest author remains static
+regardless of the eventual replacement that a build system will execute.
+
+The second mapping required at build-time, known as `image-replacements.yaml`, will take
+a similar form, mapping the common names of images to their explicitly resolved pull
+specifications. This file will have the following format:
+
+```yaml
+imagePullSpecs:
+- name: common-name
+  pullSpec: quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:298d0b496f67a0ec97a37c6e930d76db9ae69ee6838570b0cd0c688f07f53780
+```
+
+`operator-sdk generate bundle` will use this chain of mapping to perform replacements
+in the manifests before creating a bundle image layer. The layer creation will also
+be shared logic with `oc adm release new` in order to allow both processes to build
+image layers without requiring the use of a container runtime, other build system, any
+elevated permissions, privileges, capacities or SELinux roles. As the output image
+layer in both cases is `FROM scratch` and simply contains manifest data, this build
+process is simple and producing the layer by creating the underlying tar bundle does
+not come with risks.
+
+The bundle creation process will commit the `image-replacements.yaml` file into the output
+bundle, and will furthermore create as output a `release-metadata.yaml` metadata file
+that will expose any further build-time inputs used to create the bundle, so that downstream
+consumers can access this data as necessary. The format for the `release-metadata.yaml` file
+is a loose key-value store:
+
+```yaml
+metadata:
+  some-key: some-value
+```
+
+The `image-replacments.yaml` content will also be injected into the CSV using the extant
+`relatedImages` stanza for backwards compatibility, but this data is not expected to be
```

Comment:
Yes, populating `relatedImages` automatically is a great direction.

Could you clarify:
1. What if `relatedImages` is already defined in the CSV in source control? Is it overwritten, merged, or is an error raised?
1. If the answer to the above is yes, will image replacements in a pre-existing `relatedImages` block be supported?
1. There could be cases where populating `relatedImages` is not desirable, but image replacements are. Would there be an option to turn off populating the `relatedImages` block?

Perhaps this is too low level for this design.

Reply:
1. The current implementation of `relatedImages` ([doc](https://github.com/operator-framework/operator-lifecycle-manager/blob/4edf102c1054b9e7e8de772f9d0316425cca6513/doc/contributors/design-proposals/related-images.md#associating-related-images)) provides a mechanism for OLM to know which images must be mirrored when users install into a disconnected environment, that is all. That mechanism is slightly related to the current proposal but ultimately has nothing to do with it. I don't see why a user would _want_ to provide any existing `relatedImages` if they can be handled automatically by the tooling, so I would propose disallowing a user to set `relatedImages` explicitly.
2. Image replacements in an existing `relatedImages` are not a thing - there is no `operator-sdk` tooling that does this. I don't follow the question.
3. I don't understand how this case can be possible - if the operator deployment needs an image to exist for it to work, it must exist in `relatedImages` for this deployment to function in a disconnected environment, and not adding that image to `relatedImages` would not allow that. Could you please explain the use-case you are envisioning?

Reply:
@lcarva if you are talking in 2 about the current functionality for `containerbuildsystem/operator-manifest` this conversation (and that about forgetting a field) will be solved in the glue logic that adaps current CSV input from users in `dist-git` to use this mechanism.

Reply:
1. Raising an error if `relatedImages` is already defined in the CSV in source control sounds like a reasonable option here. Could you please update the doc to explicitly state this?
1. Based on your answer to 1, question 2 is no applicable. Let's disregard it. 
1. The use case I had in mind is if disconnected deployments are not yet supported for a given operator. I'm not entirely sure what it means to actually support disconnected deployments. Is it just populating `relatedImages`? Or is there additional work, e.g. testing, that needs to be done?

Reply:
ACK - will update for 1 & 2.

For 3 - just checked and I *think* it's in CVP somewhere that you are able to mark your operator supported in a disconnected manner, and that triggers a check that `spec.relatedImages` is set. So setting it regardless should be safe.